[AutoDiff] Generalize handling of semantic result parameters

asl opened the pull request:
Introduce "semantic result parameters" abstraction and treat inouts via this abstraction only.
Do not consider non-wrt semantic result parameters as semantic results

Fixes https://github.com/apple/swift/issues/67174

asl committed:
[NFC][AutoDiff] Introduce the notion of "semantic result parameter".
Handle differentiation of inouts via semantic result parameter abstraction.

asl committed:
Do not consider non-wrt semantic result parameters as semantic results

Fixes #67174

asl commented:
@swift-ci please test

asl commented:
@rxwei gentle ping :)

rxwei commented on include/swift/AST/Types.h:
```diff
@@ -3177,6 +3177,12 @@ class AnyFunctionType : public TypeBase {
     bool isNoDerivative() const { return Flags.isNoDerivative(); }
 
+    /// Whether the parameter might be a semantic result. This includes
+    /// inout parameters
```
```suggestion
    /// inout parameters.
```

rxwei commented on include/swift/AST/Types.h:
```diff
@@ -3177,6 +3177,12 @@ class AnyFunctionType : public TypeBase {
 
+    /// Whether the parameter might be a semantic result. This includes
+    /// inout parameters
+    bool isSemanticResult() const {
```
"Semantic result" is an AutoDiff term, isn't? Should it be named `isAutoDiffSemanticResult`?

asl replied:
Looks reasonable, yes.

asl committed:
Fix typos

Co-authored-by: Richard Wei <rxwei@apple.com>

asl committed:
Do not use all result indices as some semantic result parameters might be non-wrt and not differentiable at all

asl committed:
Fix handling of semantic result parameters for subset parameters thunk.

asl commented:
@rxwei I renamed the methods outside of autodiff code. Also added 2 related commits from WIP class differentiation branch that are relevant for semantic result parameters. PTAL

Thanks!

asl commented:
@swift-ci please test
